### ISSUE TITLE
JBoss EAP module clarification

### ISSUE THREAD

**Issue by johnament** [elastic/elasticsearch#11731]
Hi,  was hoping to get some clarification around the JBoss module.  I recently started using scripts and template searches in my app.  To do taht, I added an ASM dependency and brought it into the ES module.  However, whenever my tests run I end up getting this in the console

org.objectweb.asm.commons.Method from [Module "org.elasticsearch:main" from local module loader

This seems to be because JBoss ships with an older ASM than what ES is using.  Anything special to do here?

**Comment by dadoonet** [comment]
Until 2.0, we were shading libs so I guess you should not see such a conflict.

Just double checking, I guess you already read this, right? https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/_deploying_in_jboss_eap6_module.html

**Comment by johnament** [comment]
Yes, I saw that.
The problem is that this page (which looks like its based on old ticket I tracked a while ago w/ the module def) explicitly excludes ASM.  JBoss, even Wildfly 9, still ship ASM 3.1 and it looks like lucene needs a newer version.  I don't believe you guys are shading this dependency.

Great to hear what sounds like you're not shading in 2.0.  Should reduce some of the bloat.

**Comment by lukas-vlcek** [comment]
@dadoonet David, do you think you can point me to some ticket/commit which is relevant to un-shading in 2.x please? I would like to learn more about the background of such change.

**Comment by dadoonet** [comment]
@lukas-vlcek I think this commit is relevant: https://github.com/elastic/elasticsearch/commit/4c981ff4bfc250080d521af105b5e8589c9fc517

**Comment by dadoonet** [comment]
@johnament You're right. ASM was not shaded. I wonder if the documentation about about EAP deployment should include the right version of ASM lib so the right version will be picked up by the classloader first instead of using the one provided by JBoss.

That said, I don't know JBoss EAP. Is it super different than JBoss OSS version? 
(note: last time I deployed JBoss, it was a version 5. Things have changed in the meant time I guess :) )

**Comment by lukas-vlcek** [comment]
@dadoonet thanks.

**Comment by s1monw** [comment]
@lukas-vlcek the PR is here https://github.com/elastic/elasticsearch/pull/11522
We still provide a shaded verison of the JAR so you should still be able to benefit from this?

**Comment by lukas-vlcek** [comment]
@s1monw thanks. I welcome non-shaded packaging (building ES RPM will be easier).

**Comment by s1monw** [comment]
@lukas-vlcek  please report any issues if they come up! thanks!

**Comment by lukas-vlcek** [comment]
@johnament FYI, if you are interested in JBoss EAP 6.x BOMs they can be found here: http://maven.repository.redhat.com/techpreview/all/org/jboss/bom/eap6-supported-artifacts/
For example [EAP 6.4.1.GA](http://maven.repository.redhat.com/techpreview/all/org/jboss/bom/eap6-supported-artifacts/6.4.1.GA/eap6-supported-artifacts-6.4.1.GA.pom) uses ASM 3.3.1.

Also, may be you can use ES directly as part of your web app (WAR, EAR, ...) as opposed to EAP module?

**Comment by johnament** [comment]
@lukas-vlcek I don't think that's the question here.  It looks like Lucene 4.10 requires ASM 4.1 it looks like, as a result I had pull that in to the module itself (to avoid leaking).

**Comment by clintongormley** [comment]
I'm not a Java guy, so I can't tell or not: is this issue resolved?

**Comment by dadoonet** [comment]
@clintongormley I think it's not really solved as we may have at least to update our documentation. Need to understand what we have to update though. :)

Marking as doc for now and assigning it to me.

**Comment by dadoonet** [comment]
TBH I don't know what we should fix for 2.0. I have no experience with JBoss EAP. @lukas-vlcek Do you have any idea on how we should update the [doc](https://github.com/elastic/elasticsearch/blob/master/docs/java-api/index.asciidoc)?

**Comment by lukas-vlcek** [comment]
@dadoonet In our team we do not use ES as EAP module. We start embedded ES node to connect to ES cluster. So I am not able to tell you from my experience how to best configure EAP module for ES.

**Comment by johnament** [comment]
This is what ended up working for us.  Note that we were already packaging groovy in our dist so it was a new reference here instead of embedded.  On the flip side, we've also stopped using groovy due to the CVE's published, and all of our reports are now just expression based.

```
<module xmlns="urn:jboss:module:1.0" name="org.elasticsearch">
  <resources>
    <resource-root path="elasticsearch-${es.version}.jar"/>
      <resource-root path="elasticsearch-cloud-aws-${es.aws.version}.jar"/>
      <resource-root path="lucene-core-${lucene.version}.jar"/>
      <resource-root path="lucene-analyzers-common-${lucene.version}.jar"/>
      <resource-root path="lucene-codecs-${lucene.version}.jar"/>
      <resource-root path="lucene-expressions-${lucene.version}.jar"/>
      <resource-root path="lucene-queries-${lucene.version}.jar"/>
      <resource-root path="lucene-memory-${lucene.version}.jar"/>
      <resource-root path="lucene-highlighter-${lucene.version}.jar"/>
      <resource-root path="lucene-queryparser-${lucene.version}.jar"/>
      <resource-root path="lucene-sandbox-${lucene.version}.jar"/>
      <resource-root path="lucene-suggest-${lucene.version}.jar"/>
      <resource-root path="lucene-misc-${lucene.version}.jar"/>
      <resource-root path="lucene-join-${lucene.version}.jar"/>
      <resource-root path="lucene-grouping-${lucene.version}.jar"/>
      <resource-root path="lucene-spatial-${lucene.version}.jar"/>
      <resource-root path="antlr-runtime-3.5.jar"/>
      <resource-root path="asm-4.1.jar"/>
      <resource-root path="asm-commons-4.1.jar"/>
      <resource-root path="asm-tree-4.1.jar"/>
  </resources>
  <dependencies>
    <module name="sun.jdk" export="true" >
      <imports>
        <include path="sun/misc/Unsafe" />
      </imports>
    </module>
    <module name="org.apache.log4j"/>
    <module name="org.apache.commons.logging"/>
    <module name="javax.api"/>
    <module name="com.spatial4j"/>
    <module name="com.amazonaws"/>
    <module name="org.apache.commons.codec"/>
    <module name="org.codehaus.groovy"/>
  </dependencies>
</module>
```

This does bundle asm 4.1 directly w/ ES for classloader isolation, but the same could be done with a slot.  I did it this way to ensure that no one else accidentally uses ASM 4.1.

**Comment by dadoonet** [comment]
Thanks @lukas-vlcek 

Thanks @johnament. I guess `<module name="com.amazonaws"/>` is not needed for all, right?
I guess this was for elasticsearch 1.x? Any chance you tested a similar configuration for 2.0?

**Comment by dadoonet** [comment]
@johnament Do you want to come with documentation pull request about this? I have no idea of what is needed to be done here.

Closing but feel free to either:

* comment on this ticket how documentation needs to be fixed in the elasticsearch 2.0 context
* open a documentation PR which fix the doc (even better :p )

Thanks!
